### PR TITLE
xds: introduce versioned subscription test harness

### PR DESCRIPTION
**Pull request by Shikugawa**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: A part of https://github.com/envoyproxy/envoy/issues/10849. To introduce versioned xDS subscription test, we add `VersionedSubscriptionTestHarness`, and applied it to `DeltaSubscriptionTestHarness` as a part of resolving this.
Additional Description: 
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by Shikugawa** [comment]
Would you mind to unassign irrelevant reviewers?

File: test/common/config/subscription_test_harness.h [hunk 1]
```diff
@@ -120,5 +120,18 @@ ACTION_P(ThrowOnRejectedConfig, accept) {
   }
 }
 
+class VersionedSubscriptionTestHarness
+    : public virtual SubscriptionTestHarness,
+      public testing::TestWithParam<std::tuple<envoy::config::core::v3::ApiVersion>> {
+public:
+  bool isMajorVersion() {
```
Comment (htuch) [inline test/common/config/subscription_test_harness.h:127]: Please rename to `isV3MajorVersion()`.

File: test/common/config/delta_subscription_test_harness.h [hunk 2]
```diff
@@ -133,7 +147,7 @@ class DeltaSubscriptionTestHarness : public SubscriptionTestHarness {
     last_response_nonce_ = std::to_string(HashUtil::xxHash64(version));
     response->set_nonce(last_response_nonce_);
     response->set_system_version_info(version);
-    response->set_type_url(Config::TypeUrl::get().ClusterLoadAssignment);
+    response->set_type_url(type_url_);
```
Comment (htuch) [inline test/common/config/delta_subscription_test_harness.h:150]: I can see the point of testing different transport versions in these tests, but I don't think we need to vary the resource version.

File: test/common/config/subscription_test_harness.h (hunk 1 above)
Comment (Shikugawa) [reply test/common/config/subscription_test_harness.h]: done

File: test/common/config/delta_subscription_test_harness.h (hunk 2 above)
Comment (Shikugawa) [reply test/common/config/delta_subscription_test_harness.h]: only to use `SUPPORTED_MAJOR_API_VERSION` for now.

File: test/test_common/utility.h [hunk 3]
```diff
@@ -1135,4 +1135,13 @@ MATCHER_P(JsonStringEq, expected, "") {
   return equal;
 }
 
+#define SUPPORTED_API_VERSIONS                                                                     \
```
Comment (htuch) [inline test/test_common/utility.h:1138]: Which API version? Transport or resource?

File: test/common/config/delta_subscription_test_harness.h [hunk 4]
```diff
@@ -210,6 +232,9 @@ class DeltaSubscriptionTestHarness : public SubscriptionTestHarness {
   std::queue<std::string> nonce_acks_required_;
   std::queue<std::string> nonce_acks_sent_;
   bool subscription_started_{};
+  const std::string cluster_load_assignment_type_url_{
+      Config::getTypeUrl<envoy::config::endpoint::v3::ClusterLoadAssignment>(
```
Comment (htuch) [inline test/common/config/delta_subscription_test_harness.h:236]: I'm still confused here. We are using the same parameter for resource and transport version, and later we are expecting that we see a v3 delta discovery request only? Can you clarify the goal of what the PR is doing? Is it intended to validate v2 vs. v3 transport? Is this a stepping stone towards adding more testing along versioning lines?

Comment (Shikugawa) [reply test/common/config/delta_subscription_test_harness.h:236]: In general, we only to take care about resource API version and determine transport API version as V3 in this PR. The objective this PR is to bring seamless switching knob when we are available V4, V5.... resource API version in the future.

Comment (htuch) [reply test/common/config/delta_subscription_test_harness.h:236]: Why do we care about resource API version at all in this test? This is a test that is exercising the xDS transport, the payload shouldn't matter at all (that's why we use `ClusterLoadAssignment`, we could also have used some random unversioned resource, e.g. https://github.com/envoyproxy/envoy/blob/master/test/proto/helloworld.proto#L54

**Pull request closed without merging**